### ISSUE TITLE
spin build not work

### ISSUE THREAD

**Issue by MominIqbal-1234** [fermyon/spin-python-sdk#87]
Building component my-first-app with `componentize-py -w spin-http componentize app -o app.wasm`
Traceback (most recent call last):
  File "/home/myuser/.local/bin/componentize-py", line 8, in <module>
    sys.exit(script())
AssertionError: failed to read file "wit"

Caused by:
    No such file or directory (os error 2)
Error: Build command for component my-first-app failed with status Exited(1)

**Comment by karthik2804** [comment]
Hi @MominIqbal-1234, the error you are seeing is likely due to not having the `spin-sdk` package installed. You can install the `spin-sdk` package with the following command:

```bash
pip3 install spin-sdk
```

Let me know if that solves the issue.

**Comment by MominIqbal-1234** [comment]
i am follow this doc 
https://developer.fermyon.com/spin/v2/python-components

but error are same not solved

**Comment by karthik2804** [comment]
Could you run the following commands to check where the `site-packages` on your system is?

```bash
python3 -c 'import site; print(site.getsitepackages()[0])'
```

To check if the spin_sdk is installed could you do the following 

```bash
ls $(python3 -c 'import site; print(site.getsitepackages()[0])') | grep spin
```

**Comment by MominIqbal-1234** [comment]
thanks so much for your support problem is solved

I run all the commands without a virtual environment
but I don't understand why its not work without  virtual environment

and can you explain how to deploy .wasm  file own server

**Comment by agardnerIT** [comment]
> and can you explain how to deploy .wasm file own server

I'm still learning, so take this with a pinch of salt, but my understanding is that a `.wasm` file is a compiled bytecode file. Similar to what you'd get in Java if you compiled a `.java` (source code) file and got a `.class` file out the other side. `.class == .wasm`.

So you'll need something run interpret and run that `.wasm` file. There are many options:

- Fermyon cloud
- https://docs.docker.com/desktop/wasm/
- [Kubernetes](https://developer.fermyon.com/spin/v2/kubernetes)
- [Javascript](https://developer.mozilla.org/en-US/docs/WebAssembly/Loading_and_running#using_fetch)

**Comment by radu-matei** [comment]
@krzysztofwos did you run the commands inside a virtual environment?

```
$ python3 -m venv venv
$ source venv/bin/activate

$ spin build
```

@dicej should have more context on why that is required.

**Comment by radu-matei** [comment]
Sorry, I missed the `pip3 install -r requirements.txt` step in the environment:

```
$ python3 -m venv venv
$ source venv/bin/activate

$ pip3 install -r requirements.txt
Collecting spin-sdk==3.0.0 (from -r requirements.txt (line 1))
  Obtaining dependency information for spin-sdk==3.0.0 from https://files.pythonhosted.org/packages/7d/1d/455abbc9702557b51c7317d387b158c9b937ef2c1b5b15731bff6bf9e3d6/spin_sdk-3.0.0-py3-none-any.whl.metadata
  Using cached spin_sdk-3.0.0-py3-none-any.whl.metadata (16 kB)
Collecting componentize-py==0.13.1 (from -r requirements.txt (line 2))
  Obtaining dependency information for componentize-py==0.13.1 from https://files.pythonhosted.org/packages/18/9a/a8afa3677a0035da964868a431d3fb6cfcc9e1028f4ac3d462532689a0df/componentize_py-0.13.1-cp37-abi3-manylinux_2_28_aarch64.whl.metadata
  Using cached componentize_py-0.13.1-cp37-abi3-manylinux_2_28_aarch64.whl.metadata (3.4 kB)
Using cached spin_sdk-3.0.0-py3-none-any.whl (94 kB)
Using cached componentize_py-0.13.1-cp37-abi3-manylinux_2_28_aarch64.whl (40.2 MB)
Installing collected packages: spin-sdk, componentize-py
Successfully installed componentize-py-0.13.1 spin-sdk-3.0.0

$ spin build
```

Let me know if that makes a difference.

**Comment by radu-matei** [comment]
Yeah, that's the same thing I'm seeing as well. Paging @dicej to shed some light on what's happening.

**Comment by dicej** [comment]
The `spin-sdk` package contains its own directory of WIT files and a `componentize-py.toml` file which indicates where that directory is.  `componentize-py` searches for dependencies (including any SDK dependencies with `componentize-py.toml` and WIT files) by traversing any directories explicitly specified via one or more `-p` options.  In addition, it tries to guess where dependencies are stored using these heuristics: https://github.com/bytecodealliance/componentize-py/blob/35bf29e18b2c1fbaea371903a7d19e4f99b85d2b/src/command.rs#L171-L226

I.e. it first checks if the `VIRTUAL_ENV` environment variable is set (due to an activated virtual environment).  If that's not set, it tries using `pipenv --venv`.  If that doesn't work, it uses `python3 -c import site; print(site.getsitepackages()[0])`.  If none of those work, and no `-p` options were specified, it uses only the current directory.

So in order to debug why `spin-sdk` can't be found outside of a virtual environment, the first thing to do is try running `python3 -c import site; print(site.getsitepackages()[0])`.  If that fails or returns an empty string, or if it returns a directory that doesn't contain `spin-sdk`, then that will explain why it's not working.  In that case, hopefully someone with more extensive Python knowledge can tell me the right way to find the directory that has system-wide installed packages like `spin-sdk` :)

**Comment by dicej** [comment]
One possible explanation: `site.getsitepackages()[0]` only gets the first directory, so if there is more than one and `spin-sdk` is in one of the others, `componentize-py` won't find it.  It should probably print all the directories and add them all to the search path.  We can check if that's the problem by running `python3 -c import site; print(site.getsitepackages())` and seeing if that prints more than one directory (and if `spin-sdk` is in one that's not the first).

**Comment by dicej** [comment]
BTW, happy to debug this myself if someone can provide a way to reproduce it, e.g. via a dev container.

**Comment by dicej** [comment]
https://github.com/bytecodealliance/componentize-py/pull/81

**Comment by dicej** [comment]
OK, I just released `componentize-py` 0.13.2 and tested it locally.  Please upgrade to that (e.g. `pip install --upgrade componentize-py`) and let me know how it goes.

**Comment by dicej** [comment]
Sounds like it's fixed.  Please reopen if this pops up again.
